### PR TITLE
Refactor Dockerfile and Makefile

### PR DESCRIPTION
**Pull request by lucacome**
### Proposed changes
- Use only one Dockerfile (for almost all images) so we can reuse layers.
- Different make targets for the images
- Added help to Makefile

### CODE REVIEW

**Comment by lucacome** [comment]
If this looks good the docs will be updated.

File: build/Dockerfile [hunk 1]
```diff
@@ -1,45 +1,270 @@
 ARG GOLANG_CONTAINER=golang:latest
+ARG NGINX_VERSION
+ARG BUILD_OS
 
-FROM nginx:1.19.3 AS base
+######### Base image for Debian #########
+FROM nginx:$NGINX_VERSION AS debian
+ARG IC_VERSION
```
Comment (mikestephen) [inline build/Dockerfile:7]: Would it still work if this `ARG` was defined at the top with the others?

**Comment by mikestephen** [comment]
Overall, looks very clean indeed!

File: build/Dockerfile (hunk 1 above)
Comment (lucacome) [reply build/Dockerfile:7]: It wasn't working for some reason...I can double-check

Comment (mikestephen) [reply build/Dockerfile:8]: No problem at all.

File: Makefile [hunk 2]
```diff
@@ -1,83 +1,105 @@
...
 
 GIT_COMMIT = $(shell git rev-parse --short HEAD)
 
 export DOCKER_BUILDKIT = 1
 
-lint:
+.DEFAULT_GOAL:=help
+
+.PHONY: help
+help: ## Dispaly this help
```
Comment (Dean-Coakley) [inline Makefile:19]:
```suggestion
help: ## Display this help
```

File: Makefile [hunk 3]
```diff
@@ -1,83 +1,105 @@
...
+.PHONY: openshift-image-nap-plus
+openshift-image-nap-plus: ## Create Docker image for Ingress Controller (openshift with plus and nap)
+	docker build $(DOCKER_BUILD_OPTIONS) --target $(TARGET) --secret id=nginx-repo.crt,src=tempdir/nginx-repo.crt --secret id=nginx-repo.key,src=tempdir/nginx-repo.key --secret id=rhel_license,src=tempdir/rhel_license -f build/DockerfileWithAppProtectForPlusForOpenShift -t $(PREFIX):$(TAG) .
+
+.PHONY: debian-image-opentracing
+debian-image-opentracing: ## Create Docker image for Ingress Controller (with opentracing)
+	docker build $(DOCKER_BUILD_OPTIONS) --target $(TARGET) -f build/DockerfileWithOpentracing -t $(PREFIX):$(TAG) .
+
+.PHONY: debian-image-opentracing-plus
+debian-image-opentracing-plus: ## Create Docker image for Ingress Controller (with opentracinga and plus)
```
Comment (Dean-Coakley) [inline Makefile:87]:
```suggestion
debian-image-opentracing-plus: ## Create Docker image for Ingress Controller (with opentracing and plus)
```

File: Makefile [hunk 4]
```diff
@@ -1,83 +1,105 @@
...
-endif
 
-clean:
+.PHONY: clean
+clean:  ## Remove nginx-ingress binary and temp folder
 	rm -f nginx-ingress
 	rm -rf tempdir
+
+.PHONY: deps
+deps: ## Add missing and remove unsed modules, verify deps and make a vendored copy
```
Comment (Dean-Coakley) [inline Makefile:100]:
```suggestion
deps: ## Add missing and remove unused modules, verify deps and make a vendored copy
```

File: Makefile [hunk 5]
```diff
@@ -1,83 +1,105 @@
-all: push
-
 VERSION = edge
 TAG = $(VERSION)
 PREFIX = nginx/nginx-ingress
-
+NGINX_VERSION = 1.19.3
 GOLANG_CONTAINER = golang:1.15
```
Comment (Dean-Coakley) [inline Makefile:5]: I think the following should work, and would speed up initial builds for those without any golang docker images?
```suggestion
GOLANG_CONTAINER = golang:1.15-alpine
```



```
$ docker images golang
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
golang              1.15-alpine         1de1afaeaa9a        7 days ago          299MB
golang              1.15                75605a415539        3 months ago        839MB
```

**Review by Dean-Coakley — COMMENTED** [review]
Looks great 😍 

Just some typos and one suggestion/question.

**Review by mikestephen — APPROVED** [review]
Some really good changes here, looks much cleaner!

File: .dockerignore [hunk 6]
```diff
@@ -0,0 +1,4 @@
+deployments
+docs*
+examples*
+grafana
```
Comment (Dean-Coakley) [inline .dockerignore:4]: Should `tests`, `perf-tests` also be ignored?

**Review by Dean-Coakley — COMMENTED** [review]
Looks good to me. I think you can proceed with the docs.

File: build/Dockerfile [hunk 7]
```diff
@@ -1,45 +1,262 @@
...
-USER nginx
+EXPOSE 80 443
 
 ENTRYPOINT ["/nginx-ingress"]
 
 
-FROM base AS local
-COPY nginx-ingress /
+######### Cleanup for Debian #########
+FROM files as cleanup-debian
```
Comment (pleshakov) [inline build/Dockerfile:198]: are separate clean up stages necessary?
because we're doing clean up in a  separate layer, cleaning up won't make much difference -- because the files are already stored in the previous layer and thus removing them will not affect the size of the image

**Comment by Dean-Coakley** [comment]
Can you resolve the mixed indentation introduced in https://github.com/nginxinc/kubernetes-ingress/commit/095a7510a7ec1790b7cd69b8b3a95a72b288de81 in this PR?

**Comment by lucacome** [comment]
I think the `nginx-repo.*` are encoded right @ciarams87 ? Can we change it now that the action supports multiline secrets?

**Comment by ciarams87** [comment]
> I think the `nginx-repo.*` are encoded right @ciarams87 ? Can we change it now that the action supports multiline secrets?

Sure, @lucacome I can change the secrets. To allow for a transition though, it would probably be better to create new, unencoded secrets first, then I can change the names of the secrets back in a later PR. I'll let you know when I have that done.

Also, the same changes are going to have to be added to the nightly.yml file too.

**Review by Dean-Coakley — APPROVED** [review]
Changes look good.

Seeing as NGINX Plus supports Alpine since [R17](https://docs.nginx.com/nginx/releases/#r17) is there anything stopping us from making our Plus image Alpine based instead of Debian? - Just wondering, could be done in a separate PR.

**Review by ciarams87 — APPROVED** [review]
This is so much neater 😃  Great to see all the images are building, hopefully the smoke tests won't be too tricky to figure out!

**Comment by lucacome** [comment]
> This is so much neater 😃 Great to see all the images are building, hopefully the smoke tests won't be too tricky to figure out!

Seems like something is not working right with AP, all the other smoke tests would pass but they get canceled.

**Review by vepatel — APPROVED** [review]

File: docs-web/app-protect/installation.md [hunk 8]
```diff
@@ -10,17 +10,17 @@ You can also [install the Ingress Controller with App Protect by using Helm](/ng
...
 
-    When running the `make` command to build the image, be sure to use the `DOCKERFILE=appprotect/DockerfileWithAppProtectForPlus` build parameter. For example:
+    When running the `make` command to build the image, be sure to use the `debian-image-nap-plus` target. For example:
 
     ```bash
-    make DOCKERFILE=appprotect/DockerfileWithAppProtectForPlus PREFIX=<your Docker registry domain>/nginx-plus-ingress
+    make debian-image-nap-plus PREFIX=<your Docker registry domain>/nginx-plus-ingress
     ```
-    Alternatively, if you want to run on an [OpenShift](https://www.openshift.com/) cluster You can use the `DOCKERFILE=appprotect/DockerfileWithAppProtectForPlusForOpenShift` build parameter.
+    Alternatively, if you want to run on an [OpenShift](https://www.openshift.com/) cluster You can use the `openshift-image-nap-plus` target.
```
Comment (jputrino) [inline docs-web/app-protect/installation.md:20]:
```suggestion
    Alternatively, if you want to run on an [OpenShift](https://www.openshift.com/) cluster, you can use the `openshift-image-nap-plus` target.
```

File: docs-web/installation/building-ingress-controller-image.md [hunk 9]
```diff
@@ -40,48 +44,51 @@ We build the image using the make utility and the provided `Makefile`. Let’s c
...
+* **debian-image**: for building a debian-based image with NGINX.
+* **alpine-image**: for building an alpine-based image with NGINX.
+* **debian-image-plus**: for building an debian-based image with NGINX Plus.
+* **debian-image-opentracing**: for building a debian-based image with NGINX, [opentracing](https://github.com/opentracing-contrib/nginx-opentracing) module and the [Jaeger](https://www.jaegertracing.io/) tracer.
+* **debian-image-opentracing-plus**: for building a debian-based image with NGINX Plus, [opentracing](https://github.com/opentracing-contrib/nginx-opentracing) module and the [Jaeger](https://www.jaegertracing.io/) tracer.
+* **openshift-image**: for building an ubi-based image with NGINX for [Openshift](https://www.openshift.com/) clusters.
+* **openshift-image-plus**: for building an ubi-based image with NGINX Plus for [Openshift](https://www.openshift.com/) clusters.
+* **debian-image-nap-plus**: for building a debian-based image with NGINX Plus and the [appprotect](/nginx-app-protect/) module.
+Note: You need to place a file named `rhel_license` containing Your Organization and Activation key in the project root. Example:
```
Comment (jputrino) [inline docs-web/installation/building-ingress-controller-image.md:75]:
```suggestion
Note: You need to store your RHEL organization and activation keys in a file named `rhel_license` in the project root. Example:
```

File: docs-web/installation/building-ingress-controller-image.md [hunk 10]
```diff
@@ -40,48 +44,51 @@ We build the image using the make utility and the provided `Makefile`. Let’s c
...
 * **DOCKER_BUILD_OPTIONS** -- the [options](https://docs.docker.com/engine/reference/commandline/build/#options) for the `docker build` command. For example, `--pull`.
-* **BUILD_IN_CONTAINER** -- By default, to compile the controller we use the [golang](https://hub.docker.com/_/golang/) container that we run as part of the building process. If you want to compile the controller using your local golang environment:
-  1. Make sure that the Ingress controller repo is in your `$GOPATH`.
-  1. Specify `BUILD_IN_CONTAINER=0` when you run the make command.
+* **TARGET** -- By default, the Ingress Controller locally is compiled locally using a `local` golang environment. If you want to compile the controller using your local golang environment make sure that the Ingress controller repo is in your `$GOPATH`. To compile the Ingress Controller using Docker the [golang](https://hub.docker.com/_/golang/) container, specify `TARGET=container`.
```
Comment (jputrino) [inline docs-web/installation/building-ingress-controller-image.md:94]:
```suggestion
* **TARGET** -- By default, the Ingress Controller is compiled locally using a `local` golang environment. If you want to compile the controller using your local golang environment, make sure that the Ingress Controller repo is in your `$GOPATH`. To compile the Ingress Controller using the Docker [golang](https://hub.docker.com/_/golang/) container, specify `TARGET=container`.
```

File: docs-web/installation/building-ingress-controller-image.md [hunk 11]
```diff
@@ -40,48 +44,51 @@ We build the image using the make utility and the provided `Makefile`. Let’s c
...
+  ```bash
+  RHEL_ORGANIZATION=1111111
+  RHEL_ACTIVATION_KEY=your-key
+  ```
+
+A few other useful targets:
+* **push**: pushes the image to the Docker registry specified in `PREFIX` and `TAG` variables.
+* **all**: executes test `test`, `lint`, `verify-codegen`, `update-crds` and `debian-image`. If one of the targets fails, the execution process stops, reporting an error.
+* **test**: runs unit tests.
+* **certificate-and-key**: The Ingress controller requires a certificate and a key for the default HTTP/HTTPS server. You can reference them in a TLS Secret in a command-line argument to the Ingress controller. As an alternative, you can add a file in the PEM format with your certificate and key to the image as `/etc/nginx/secrets/default`. Optionally, you can generate a self-signed certificate and a key using this target. Note that you must add the `ADD` instruction in the Dockerfile to copy the cert and the key to the image.
```
Comment (jputrino) [inline docs-web/installation/building-ingress-controller-image.md:85]:
```suggestion
* **certificate-and-key**: The Ingress Controller requires a certificate and a key for the default HTTP/HTTPS server. You can reference them in a TLS Secret in a command-line argument to the Ingress Controller. As an alternative, you can add a file in the PEM format with your certificate and key to the image as `/etc/nginx/secrets/default`. Optionally, you can generate a self-signed certificate and a key using this target. Note that you must add the `ADD` instruction in the Dockerfile to copy the cert and the key to the image.
```

File: docs-web/installation/building-ingress-controller-image.md [hunk 12]
```diff
@@ -17,7 +17,7 @@ Although the Ingress controller is written in golang, golang is not required, as
 
 We build the image using the make utility and the provided `Makefile`. Let’s create the controller binary, build an image and push the image to the private registry.
 
-1. Make sure to run the `docker login` command first to login to the registry. If you’re using Google Container Registry, you don’t need to use the docker command to login -- make sure you’re logged into the gcloud tool (using the `gcloud auth login` command) and set the variable `PUSH_TO_GCR=1` when running the make command.
+1. Make sure to run the `docker login` command first to login to the registry. If you’re using Google Container Registry, you don’t need to use the docker command to login -- make sure you’re logged into the gcloud tool (using the `gcloud auth login` and `gcloud auth configure-docker` commands).
```
Comment (jputrino) [inline docs-web/installation/building-ingress-controller-image.md:20]:
```suggestion
1. Make sure to run the `docker login` command first to log in to the registry. 

   If you’re using Google Container Registry, make sure you’re logged into the gcloud tool by running the `gcloud auth login` and `gcloud auth configure-docker` commands.
```

File: docs-web/app-protect/installation.md [hunk 13]
```diff
@@ -10,17 +10,17 @@ You can also [install the Ingress Controller with App Protect by using Helm](/ng
...
 
     If you intend to use [external references](https://docs.nginx.com/nginx-app-protect/configuration/#external-references) in NGINX App Protect policies, you may want to provide a custom CA certificate to authenticate with the hosting server.
-    In order to do that, place the *.crt file in the build/appprotect folder and uncomment the lines following this comment: `#Uncomment the lines below if you want to install a custom CA certificate`
+    In order to do that, place the *.crt file in the build folder and uncomment the lines following this comment: `#Uncomment the lines below if you want to install a custom CA certificate`
```
Comment (jputrino) [inline docs-web/app-protect/installation.md:23]:
```suggestion
    In order to do so, place the `*.crt` file in the build folder and uncomment the lines that follow this comment: `#Uncomment the lines below if you want to install a custom CA certificate`
```

**Review by jputrino — APPROVED** [review]
LGTM, just a few suggestions

File: build/Dockerfile [hunk 14]
```diff
@@ -1,46 +1,289 @@
...
+	cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DBUILD_STATIC_LIBS=OFF -DBUILD_MOCKTRACER=OFF -DENABLE_LINTING=OFF .. && \
+	make && make install
+
+RUN cd nginx-${NGINX_VERSION} && \
+	./configure \
+	--with-compat --with-debug \
+	--add-dynamic-module=/nginx-opentracing/opentracing && \
+	make modules
+
+FROM debian AS tracer-downloader
```
Comment (pleshakov) [inline build/Dockerfile:207]: could you add a comment here for consistency and visibility?

File: build/Dockerfile [hunk 15]
```diff
@@ -1,46 +1,289 @@
...
+
+ARG JAEGER_VERSION=v0.4.2
+
+RUN set -x \
+	&& apt-get update \
+	&& apt-get install --no-install-recommends --no-install-suggests -y ca-certificates apt-transport-https wget \
+	&& wget https://github.com/jaegertracing/jaeger-client-cpp/releases/download/${JAEGER_VERSION}/libjaegertracing_plugin.linux_amd64.so -O /usr/local/lib/libjaegertracing_plugin.so
+
+
+FROM debian as opentracing
```
Comment (pleshakov) [inline build/Dockerfile:217]: could  you add a comment here for consistency and visibility?

File: build/Dockerfile [hunk 16]
```diff
@@ -49,5 +292,14 @@ RUN CGO_ENABLED=0 GOFLAGS='-mod=vendor' \
 	go build -installsuffix cgo -ldflags "-w -X main.version=${VERSION} -X main.gitCommit=${GIT_COMMIT}" -o /nginx-ingress
 
 
-FROM base AS container
-COPY --from=builder /nginx-ingress /
+############################################# Create image with nginx-ingress built in #############################################
```
Comment (pleshakov) [inline build/Dockerfile:295]: ?
```suggestion
############################################# Create image with nginx-ingress built in container #############################################
```

File: build/Dockerfile [hunk 17]
```diff
@@ -1,46 +1,289 @@
...
+	&& echo "baseurl=https://plus-pkgs.nginx.com/centos/8/\$basearch/" >> /etc/yum.repos.d/nginx-plus-8.repo \
+	&& echo "sslclientcert=/etc/ssl/nginx/nginx-repo.crt" >> /etc/yum.repos.d/nginx-plus-8.repo \
+	&& echo "sslclientkey=/etc/ssl/nginx/nginx-repo.key" >> /etc/yum.repos.d/nginx-plus-8.repo \
+	&& echo "gpgcheck=1" >> /etc/yum.repos.d/nginx-plus-8.repo \
+	&& echo "enabled=1" >> /etc/yum.repos.d/nginx-plus-8.repo
+
+# Install NGINX Plus
+RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode=0644 \
+	--mount=type=secret,id=nginx-repo.key,dst=/etc/ssl/nginx/nginx-repo.key,mode=0644 \
+	yum install -y ca-certificates nginx-plus-${NGINX_PLUS_VERSION}
```
Comment (pleshakov) [inline build/Dockerfile:159]: do we need to do clean up like here https://github.com/nginxinc/kubernetes-ingress/blob/master/build/openshift/DockerfileForPlus#L50 ?

File: build/Dockerfile [hunk 18]
```diff
@@ -1,46 +1,289 @@
...
+FROM openshift-base AS openshift
+ARG NGINX_VERSION
+
+RUN echo "[nginx]" >> /etc/yum.repos.d/nginx.repo \
+	&& echo "name=nginx repo" >> /etc/yum.repos.d/nginx.repo \
+	&& echo "baseurl=https://nginx.org/packages/mainline/rhel/8/\$basearch/" >> /etc/yum.repos.d/nginx.repo \
+	&& echo "gpgcheck=0" >> /etc/yum.repos.d/nginx.repo \
+	&& echo "enabled=1" >> /etc/yum.repos.d/nginx.repo \
+	&& echo "module_hotfixes=true" >> /etc/yum.repos.d/nginx.repo \
+	&& yum install -y nginx-${NGINX_VERSION}
```
Comment (pleshakov) [inline build/Dockerfile:138]: do we need to do clean up like https://github.com/nginxinc/kubernetes-ingress/blob/master/build/openshift/Dockerfile#L32 ?

File: build/Dockerfile [hunk 19]
```diff
@@ -1,46 +1,289 @@
...
+	&& gpg --no-default-keyring --keyring app_protect_keyring.gpg --export > /etc/apt/trusted.gpg.d/nginx_app_signing.gpg \
+	&& sed -i '$ d' /etc/apt/apt.conf.d/90nginx \
+	&& echo "Acquire::https::plus-pkgs.nginx.com::User-Agent  \"k8s-ic-$IC_VERSION-app-$APPPROTECT_VERSION-apt\";" >> /etc/apt/apt.conf.d/90nginx \
+	&& echo "deb https://app-protect-sigs.nginx.com/debian/ buster nginx-plus\n" | tee /etc/apt/sources.list.d/app-protect-sigs.list \
+	&& echo "Acquire::https::app-protect-sigs.nginx.com::Verify-Peer \"true\";" >> /etc/apt/apt.conf.d/90app-protect-sigs \
+	&& echo "Acquire::https::app-protect-sigs.nginx.com::Verify-Host \"true\";" >> /etc/apt/apt.conf.d/90app-protect-sigs \
+	&& echo "Acquire::https::app-protect-sigs.nginx.com::SslCert     \"/etc/ssl/nginx/nginx-repo.crt\";" >> /etc/apt/apt.conf.d/90app-protect-sigs \
+	&& echo "Acquire::https::app-protect-sigs.nginx.com::SslKey      \"/etc/ssl/nginx/nginx-repo.key\";" >> /etc/apt/apt.conf.d/90app-protect-sigs \
+	&& apt-get purge --auto-remove -y gnupg wget \
+	&& rm -rf /var/lib/apt/lists/*
```
Comment (pleshakov) [inline build/Dockerfile:90]: what about files in /etc/apt/apt.conf.d like https://github.com/nginxinc/kubernetes-ingress/blob/master/build/appprotect/DockerfileWithAppProtectForPlus ?

File: build/Dockerfile [hunk 20]
```diff
@@ -1,46 +1,289 @@
...
+	--mount=type=secret,id=nginx-repo.key,dst=/etc/ssl/nginx/nginx-repo.key,mode=0644 \
+	apt-get update && apt-get install -y apt-transport-https nginx-plus-module-appprotect=$APPPROTECT_MODULE_VERSION \
+	app-protect-plugin=$APPPROTECT_PLUGIN_VERSION \
+	app-protect-engine=$APPPROTECT_ENGINE_VERSION \
+	app-protect-compiler=$APPPROTECT_COMPILER_VERSION \
+	app-protect=$APPPROTECT_MODULE_VERSION \
+	&& apt-get install -y app-protect-attack-signatures${APPPROTECT_SIG_VERSION:+=$APPPROTECT_SIG_VERSION} \
+	&& apt-get install -y app-protect-threat-campaigns${APPPROTECT_THREAT_CAMPAIGNS_VERSION:+=$APPPROTECT_THREAT_CAMPAIGNS_VERSION} \
+	&& apt-get purge --auto-remove -y apt-transport-https \
+	&& rm -rf /var/lib/apt/lists/*
```
Comment (pleshakov) [inline build/Dockerfile:102]: what about files in /etc/apt/apt.conf.d like here https://github.com/nginxinc/kubernetes-ingress/blob/master/build/appprotect/DockerfileWithAppProtectForPlus#L67 ?

File: .dockerignore (hunk 6 above)
Comment (lucacome) [reply .dockerignore:3]: Adding a comment for the record, they can't be added because tests are used in docker

File: build/Dockerfile [hunk 21]
```diff
@@ -1,46 +1,289 @@
...
+	--mount=type=secret,id=nginx-repo.key,dst=/etc/ssl/nginx/nginx-repo.key,mode=0644 \
+	apt-get update && \
+	apt-get install --no-install-recommends --no-install-suggests nginx-plus-module-opentracing=${NGINX_OPENTRACING_MODULE_VERSION} && \
+	rm -rf /var/lib/apt/lists/*
+
+COPY --from=tracer-downloader /usr/local/lib/libjaegertracing_plugin.so /usr/local/lib/libjaegertracing_plugin.so
+
+############################################# Cleanup for Debian with NGINX Plus and App Protect #############################################
+FROM debian-plus-ap-intermediate as debian-plus-ap
+RUN rm /etc/apt/apt.conf.d/90app-protect-sigs /etc/apt/sources.list.d/app-protect-sigs.list \
```
Comment (pleshakov) [inline build/Dockerfile:242]: for readability, would it make sense to create each file/folder on a separate line like here https://github.com/nginxinc/kubernetes-ingress/blob/master/build/appprotect/DockerfileWithAppProtectForPlus#L78-L96 ?

File: build/Dockerfile [hunk 22]
```diff
@@ -1,46 +1,289 @@
...
+
+COPY --chown=nginx:0 build/log-default.json /etc/nginx
+
+
+############################################# Create common files, permissions and setcap #############################################
+FROM $BUILD_OS as files
+
+ARG PLUS
+
+RUN mkdir -p /var/lib/nginx /etc/nginx/secrets /etc/nginx/stream-conf.d /etc/nginx/oidc \
```
Comment (pleshakov) [inline build/Dockerfile:266]: for readability, would it make sense to create  each file on a separate line like we had before https://github.com/nginxinc/kubernetes-ingress/blob/master/build/DockerfileForPlus#L55 ?

File: build/Dockerfile [hunk 23]
```diff
@@ -1,46 +1,289 @@
...
+	&& echo "Acquire::https::plus-pkgs.nginx.com::User-Agent  \"k8s-ic-$IC_VERSION-apt\";" >> /etc/apt/apt.conf.d/90nginx \
+	&& printf "deb https://plus-pkgs.nginx.com/debian buster nginx-plus\n" > /etc/apt/sources.list.d/nginx-plus.list \
+	&& apt-get purge --auto-remove -y gnupg wget \
+	&& rm -rf /var/lib/apt/lists/*
+
+RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode=0644 \
+	--mount=type=secret,id=nginx-repo.key,dst=/etc/ssl/nginx/nginx-repo.key,mode=0644 \
+	apt-get update && apt-get install -y apt-transport-https libcap2-bin nginx-plus=${NGINX_PLUS_VERSION} nginx-plus-module-njs=${NGINX_NJS_VERSION} \
+	&& apt-get purge --auto-remove -y apt-transport-https \
+	&& rm -rf /var/lib/apt/lists/*
```
Comment (pleshakov) [inline build/Dockerfile:56]: what about removing files `/etc/apt/apt.conf.d/90nginx /etc/apt/sources.list.d/nginx-plus.list` like we do here  https://github.com/nginxinc/kubernetes-ingress/blob/master/build/DockerfileForPlus#L46 ?

File: build/Dockerfile [hunk 24]
```diff
@@ -1,46 +1,289 @@
...
+############################################# Build image for Opentracing with NGINX Plus #############################################
+FROM debian-plus as opentracing-plus
+
+ENV NGINX_OPENTRACING_MODULE_VERSION 23+0.9.0-1~buster
+
+RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode=0644 \
+	--mount=type=secret,id=nginx-repo.key,dst=/etc/ssl/nginx/nginx-repo.key,mode=0644 \
+	apt-get update && \
+	apt-get install --no-install-recommends --no-install-suggests nginx-plus-module-opentracing=${NGINX_OPENTRACING_MODULE_VERSION} && \
+	rm -rf /var/lib/apt/lists/*
```
Comment (pleshakov) [inline build/Dockerfile:236]: what about removing files in apt.conf.d like we do here https://github.com/nginxinc/kubernetes-ingress/blob/master/build/DockerfileWithOpentracingForPlus#L59 ?

File: build/Dockerfile (hunk 21 above)
Comment (pleshakov) [inline build/Dockerfile:242]: I think `rm /etc/apt/apt.conf.d/90app-protect-sigs /etc/apt/sources.list.d/app-protect-sigs.list` could be moved to the corresponding debin app protect stage, no?

File: docs-web/installation/building-ingress-controller-image.md [hunk 25]
```diff
@@ -40,48 +46,52 @@ We build the image using the make utility and the provided `Makefile`. Let’s c
...
-* **container**: builds a Docker image.
-* **push**: pushes the image to the private Docker registry.
-* **all** (the default target): executes the four targets above in the order listed. If one of the targets fails, the execution process stops, reporting an error.
+You can see a list of all the targets by running `make` without any target or `make help`
+
+Below you can find some of the most useful targets in the **Makefile**:
+* **build**: creates the Ingress Controller binary using the local golang environment (ignored when `TARGET` is `container`).
+* **debian-image**: for building a debian-based image with NGINX.
+* **alpine-image**: for building an alpine-based image with NGINX.
+* **debian-image-plus**: for building an debian-based image with NGINX Plus.
```
Comment (pleshakov) [inline docs-web/installation/building-ingress-controller-image.md:71]:
```suggestion
* **debian-image-plus**: for building a debian-based image with NGINX Plus.
```

File: tests/suite/resources_utils.py [hunk 26]
```diff
@@ -22,6 +22,7 @@ class RBACAuthorization:
         role (str): cluster role name
         binding (str): cluster role binding name
     """
+
```
Comment (pleshakov) [inline tests/suite/resources_utils.py:25]: seems like irrelevant changes for this PR

**Review by pleshakov — COMMENTED** [review]
Hi @lucacome 

Could we make the default `TARGET=container`. Otherwise, if you just follow instructions here https://deploy-preview-1241--nginx-kubernetes-ingress.netlify.app/nginx-ingress-controller/installation/building-ingress-controller-image/#building-the-image-and-pushing-it-to-the-private-registry , you will get :
```
$ make debian-image-plus PREFIX=myregistry.example.com/nginx-plus-ingress
Docker version 20.10.4, build d3cb89e
make: go: Command not found
make: *** [Makefile:51: build] Error 127
```
Note that previously we had build in container enabled as default. This is for users/customers, as we don't expect them to have the go installed/configured on their systems.

Also I added the "change" label to this PR, as it introduces two changes:
1. changes in `Makefile` - customers will need to use different command and arguments to build the images
1. changes in Dockerfiles - customers who customized their Dockerfiles will have to migrate their customizations to the new Dockerfile.

Those changes will have to be reflected in our CHANELOG. Could you update the description of this PR with upgrade notes related to the two changes? This way we will insert them into our CHANGELOG into the UPGRADE section.

Also, I left a number of additional comments

File: build/Dockerfile (hunk 23 above)
Comment (lucacome) [reply build/Dockerfile:54]: they're needed in the ap stage

File: build/Dockerfile [hunk 27]
```diff
@@ -1,46 +1,289 @@
...
+ENV NGINX_NJS_VERSION 23+0.5.0-1~buster
+
+LABEL version=debian-nginx-plus-$NGINX_PLUS_VERSION-$IC_VERSION
+LABEL maintainer="NGINX Docker Maintainers <docker-maint@nginx.com>"
+
+RUN set -x \
+	&& apt-get update \
+	&& apt-get install --no-install-recommends --no-install-suggests -y ca-certificates gnupg wget \
+	&& wget https://nginx.org/keys/nginx_signing.key \
+	&& gpg --no-default-keyring --keyring nginx_keyring.gpg --import nginx_signing.key \
```
[the remark this answers is not in the document]
Comment (lucacome) [reply build/Dockerfile:43]: yes

File: build/Dockerfile (hunk 17 above)
Comment (lucacome) [reply build/Dockerfile]: `/etc/ssl/nginx` is empty because we use the mount

File: build/Dockerfile [hunk 28]
```diff
@@ -1,46 +1,289 @@
...
+
+RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode=0644 \
+	--mount=type=secret,id=nginx-repo.key,dst=/etc/ssl/nginx/nginx-repo.key,mode=0644 \
+	apt-get update && \
+	apt-get install --no-install-recommends --no-install-suggests nginx-plus-module-opentracing=${NGINX_OPENTRACING_MODULE_VERSION} && \
+	rm -rf /var/lib/apt/lists/*
+
+COPY --from=tracer-downloader /usr/local/lib/libjaegertracing_plugin.so /usr/local/lib/libjaegertracing_plugin.so
+
+############################################# Cleanup for Debian with NGINX Plus and App Protect #############################################
```
[the remark this answers is not in the document]
Comment (lucacome) [reply build/Dockerfile]: removed

File: build/Dockerfile (hunk 22 above)
[the remark this answers is not in the document]
Comment (lucacome) [reply build/Dockerfile]: removed

File: build/Dockerfile (hunk 14 above)
Comment (lucacome) [reply build/Dockerfile:224]: added

File: build/Dockerfile (hunk 15 above)
Comment (lucacome) [reply build/Dockerfile:235]: added

**Comment by lucacome** [comment]
> Could we make the default `TARGET=container`.

> Note that previously we had build in container enabled as default. This is for users/customers, as we don't expect them to have the go installed/configured on their systems.

I've added a note in the docs and errors in the Makefile, to use `TARGET=container` if go is not installed. I don't think the default should be to build in a container (much slower) and for the devs to have to add an additional parameter every time. Also, the builds and tests all have the binary built locally, so it's more consistent imo.

**Comment by TechIsCool** [comment]
https://docs.nginx.com/nginx-ingress-controller/third-party-modules/opentracing/
https://www.nginx.com/blog/opentracing-with-nginx-ingress-controller-for-kubernetes/#Building-OpenTracing-into-the-NGINX-Ingress-Controller-Image

A couple places for Open Tracing need docs updated.

**Comment by lucacome** [comment]
Hi @TechIsCool 

thanks for pointing this out, the documentation is for https://github.com/nginxinc/kubernetes-ingress/releases/tag/v1.10.0. This change will go in v1.11.0 so it will be updated once we release that. Unfortunately, we don't have documentation for the `edge` version (master) at the moment.

I'll see if we can update that blog post once we release v1.11.0.

Thanks!
